### DISCUSSION TITLE
How to use a CSR matrix to get similarity

### DISCUSSION THREAD

**Discussion by Vikram12301** [facebookresearch/faiss#3563] [General]
import faiss
from scipy.sparse import csr_matrix

rows = [0, 1, 1, 2]
cols = [0, 1, 2, 1]
vals = [1, 2, 3, 4]
csr_mat = csr_matrix((vals, (rows, cols)), shape=(3, 3))

index = faiss.IndexFlatIP(csr_mat.shape[1])
index.add(csr_mat.indptr, csr_mat.indices, csr_mat.data)

I got this error on running the above code
TypeError: replacement_add() takes 2 positional arguments but 4 were given
How to get the right similarity?

**Comment by mdouze** [comment]
Faiss indexes dense vectors. 
There is some limited work on sparse vectors, see: 
https://github.com/facebookresearch/faiss/wiki/Case-studies#mixed-sparse-dense-clustering
